# Introduction

Being a Certified Personal Trainer, I dreamt of owning my own Gym in future. One major success of running a Gym is selecting its location. You can have a world-class Gym but if it is in an inaccessible area, the Gym will not be visited and it will hardly take off. Thus, I would like to take this opportunity to find the most suitable location to set up a Gym

# Business Problem

Select a location which is accessible to the public and lack fitness facilities in the present day.

# Data

Since Singapore is well-connected by the Mass Rapid Transit (MRT) System, one key data is a pre-consolidated locations (Longtitude & Latitude) of all MRT Stations in Singapore. This data was consolidated and made available on GitHub, link provided below. An enquiry to Foursquare will locate the nearby fitness facilities and identify which MRT Station lack fitness facilities and will be a suitable Gym location. As the Venue Category in Foursquare might refer to fitness facilites as "Climbing Gym", "Yoga Studio", etc., it will be necessary to consider them as well before analyse and concluding on the location.

https://github.com/hxchua/datadoubleconfirm/blob/master/datasets/mrtsg.csv

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

from botocore.client import Config
import ibm_boto3

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Libraries imported.


# Methodology

## Step 1: Areas that are accessible by MRT are considered accessible, thus, the locations of all MRT Stations are imported and their headers are renamed. As the consolidated list also contained the Light Rail Transit (LRT) stations, a new dataframe (sg_mrt) was defined to only values with "MRT".

In [3]:
# save the DataFrame as CSV file
df_sg.to_csv("sg_df.csv", index=False)
df_sg.columns = ['ID','Neighborhood','Number','X','Y','Latitude', 'Longitude','Line']

In [4]:
df_sg

ID                   Neighborhood Number           X           Y  \
0     12          ADMIRALTY MRT STATION   NS10  24402.1063  46918.1131   
1     16           ALJUNIED MRT STATION    EW9  33518.6049  33190.0020   
2     33         ANG MO KIO MRT STATION   NS16  29807.2655  39105.7720   
3     81              BAKAU LRT STATION    SE3  36026.0821  41113.8766   
4     80            BANGKIT LRT STATION    BP9  21248.2460  40220.9693   
5    153            BARTLEY MRT STATION   CC12  33168.3039  36108.7003   
6    115           BAYFRONT MRT STATION   DT16  30867.0093  29368.6250   
7    115           BAYFRONT MRT STATION    CE1  30867.0093  29368.6250   
8    140       BEAUTY WORLD MRT STATION    DT5  21598.1665  35931.2359   
9     37              BEDOK MRT STATION    EW5  38757.9520  34024.7048   
10   143        BEDOK NORTH MRT STATION   DT29  37421.7438  35214.6804   
11   149    BEDOK RESERVOIR MRT STATION   DT30  39008.3132  35421.0409   
12   100          BENCOOLEN MRT STATION   DT21  29899.2575  31247.3368   
13    99          BENDEMEER MRT STATION   DT23  31300.8084  32884.7382   
14   109             BISHAN MRT STATION   CC15  29762.3507  37046.3699   
15   158             BISHAN MRT STATION   NS17  29649.8988  36994.4330   
16   178          BOON KENG MRT STATION    NE9  31157.0712  33517.6089   
17    34           BOON LAY MRT STATION   EW27  13837.9827  35641.9191   
18   159    BOTANIC GARDENS MRT STATION    DT9  26087.2663  33852.4484   
19   185    BOTANIC GARDENS MRT STATION   CC19  25959.4046  33817.7525   
20   182           BRADDELL MRT STATION   NS18  29500.0972  35848.0687   
21    93         BRAS BASAH MRT STATION    CC2  29930.7674  31025.9016   
22   183           BUANGKOK MRT STATION   NE15  34655.3132  40537.2069   
23   103              BUGIS MRT STATION   DT14  30620.8769  31323.2495   
24   111              BUGIS MRT STATION   EW12  30491.5648  31424.3503   
25    32        BUKIT BATOK MRT STATION    NS2  18679.3541  36795.0141   
26   117        BUKIT BROWN MRT STATION   CC18  27707.4166  35102.4802   
27    35       BUKIT GOMBAK MRT STATION    NS3  18938.3305  37870.6512   
28    47      BUKIT PANJANG LRT STATION    BP6  20175.9353  39987.9068   
29   171      BUKIT PANJANG MRT STATION    DT1  20011.3981  40108.6838   
30    28        BUONA VISTA MRT STATION   EW21  23200.3600  32167.2655   
31   124        BUONA VISTA MRT STATION   CC22  23251.7611  32090.7692   
32   127          CALDECOTT MRT STATION   CC17  28691.2451  35538.7680   
33   172             CASHEW MRT STATION    DT2  20362.9591  39043.5835   
34   180     CHANGI AIRPORT MRT STATION    CG2  45254.8614  37711.0150   
35    70          CHENG LIM LRT STATION    SW1  34730.4628  42018.8886   
36   131          CHINATOWN MRT STATION    NE4  29124.9585  29643.4786   
37   132          CHINATOWN MRT STATION   DT19  29238.9683  29686.5359   
38     2     CHINESE GARDEN MRT STATION   EW25  16790.7466  36056.3019   
39    65      CHOA CHU KANG LRT STATION    BP1  18121.6052  40753.8693   
40    19      CHOA CHU KANG MRT STATION    NS4  18097.9271  40811.9589   
41    95          CITY HALL MRT STATION   EW13  30144.2851  30591.8478   
42    95          CITY HALL MRT STATION   NS25  30144.2851  30591.8478   
43   157        CLARKE QUAY MRT STATION    NE5  29473.1249  30088.7038   
44    15           CLEMENTI MRT STATION   EW23  20430.0600  33026.6075   
45     8       COMMONWEALTH MRT STATION   EW20  24103.2523  31642.6034   
46    83        COMPASSVALE LRT STATION    SE1  35475.5631  41821.5783   
47    55         CORAL EDGE LRT STATION    PE3  36820.8455  41757.0642   
48    59               COVE LRT STATION    PE1  36084.2293  42351.1334   
49   154             DAKOTA MRT STATION    CC8  34159.2741  32299.8905   
50    58              DAMAI LRT STATION    PE7  36378.1828  43009.3794   
51   144        DHOBY GHAUT MRT STATION   NS24  29424.1448  31229.3143   
52   164        DHOBY GHAUT MRT STATION    NE6  29354.3681  31340.3479   
53   165  

In [5]:
MRT_names = list(df_sg.Neighborhood.unique())

sg_with_MRT = []

for x in MRT_names:
    if "MRT" in x.upper():
        sg_with_MRT.append(x)
        
sg_with_MRT

sg_mrt = df_sg[df_sg['Neighborhood'].isin(sg_with_MRT)].reset_index(drop=True)
sg_mrt.head()

ID            Neighborhood Number           X           Y  Latitude  \
0   12   ADMIRALTY MRT STATION   NS10  24402.1063  46918.1131  1.440585   
1   16    ALJUNIED MRT STATION    EW9  33518.6049  33190.0020  1.316433   
2   33  ANG MO KIO MRT STATION   NS16  29807.2655  39105.7720  1.369933   
3  153     BARTLEY MRT STATION   CC12  33168.3039  36108.7003  1.342828   
4  115    BAYFRONT MRT STATION   DT16  30867.0093  29368.6250  1.281874   

    Longitude    Line  
0  103.800998     RED  
1  103.882893   GREEN  
2  103.849553     RED  
3  103.879746  YELLOW  
4  103.859073    BLUE

In [6]:
address = 'Singapore'

geolocator = Nominatim(user_agent="my-application")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Singapore are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Singapore are 1.357107, 103.8194992.


## Step 2: The Map is plotted based on the Latitide and Longitude above to ensure all MRT stations are represented in the data.

In [7]:
map_sg = folium.Map(location=[latitude, longitude],zoom_start=12)

# add markers to map
for lat, lng, neighborhood in zip(sg_mrt['Latitude'], sg_mrt['Longitude'], sg_mrt['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_sg) 

map_sg

In [8]:
# save the map as HTML file
map_sg.save('map_sg.html')

## Step 3: Using the Latitude and Longitude of the MRT Stations, data of the facilities nearby are retrieved from Foursquare.

In [10]:
radius = 2000
LIMIT = 100

venues = []

for lat, long, neighborhood in zip(sg_mrt['Latitude'], sg_mrt['Longitude'], sg_mrt['Neighborhood']):
    
    # create the API request URL
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    # make the GET request
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    # return only relevant information for each nearby venue
    for venue in results:
        venues.append((
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [38]:
# convert the venues list into a new DataFrame
venues_df = pd.DataFrame(venues)

# define the column names
venues_df.columns = ['Neighborhood', 'Latitude', 'Longitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']

print(venues_df.shape)
venues_df.head()

(13594, 7)


Neighborhood  Latitude   Longitude  \
0  ADMIRALTY MRT STATION  1.440585  103.800998   
1  ADMIRALTY MRT STATION  1.440585  103.800998   
2  ADMIRALTY MRT STATION  1.440585  103.800998   
3  ADMIRALTY MRT STATION  1.440585  103.800998   
4  ADMIRALTY MRT STATION  1.440585  103.800998   

                         VenueName  VenueLatitude  VenueLongitude  \
0  Kampung Admiralty Hawker Centre       1.439939      103.800774   
1         Superpets Warehouse / HQ       1.436848      103.806059   
2                       McDonald's       1.445931      103.798101   
3                       Asia Ghani       1.437431      103.795388   
4                        Starbucks       1.439761      103.800659   

          VenueCategory  
0            Food Court  
1             Pet Store  
2  Fast Food Restaurant  
3      Malay Restaurant  
4           Coffee Shop

In [12]:
venues_df.groupby(["Neighborhood"]).count()

Latitude  Longitude  VenueName  VenueLatitude  \
Neighborhood                                                                   
ADMIRALTY MRT STATION                81         81         81             81   
ALJUNIED MRT STATION                100        100        100            100   
ANG MO KIO MRT STATION              100        100        100            100   
BARTLEY MRT STATION                 100        100        100            100   
BAYFRONT MRT STATION                200        200        200            200   
BEAUTY WORLD MRT STATION             86         86         86             86   
BEDOK MRT STATION                   100        100        100            100   
BEDOK NORTH MRT STATION             100        100        100            100   
BEDOK RESERVOIR MRT STATION         100        100        100            100   
BENCOOLEN MRT STATION               100        100        100            100   
BENDEMEER MRT STATION               100        100        100            100   
BISHAN MRT STATION                  200        200        200            200   
BOON KENG MRT STATION               100        100        100            100   
BOON LAY MRT STATION                 96         96         96             96   
BOTANIC GARDENS MRT STATION         200        200        200            200   
BRADDELL MRT STATION                100        100        100            100   
BRAS BASAH MRT STATION              100        100        100            100   
BUANGKOK MRT STATION                100        100        100            100   
BUGIS MRT STATION                   200        200        200            200   
BUKIT BATOK MRT STATION             100        100        100            100   
BUKIT BROWN MRT STATION             100        100        100            100   
BUKIT GOMBAK MRT STATION             99         99         99             99   
BUKIT PANJANG MRT STATION            97         97         97             97   
BUONA VISTA MRT STATION             200        200        200            200   
CALDECOTT MRT STATION               100        100        100            100   
CASHEW MRT STATION                  100        100        100            100   
CHANGI AIRPORT MRT STATION           69         69         69             69   
CHINATOWN MRT STATION               200        200        200            200   
CHINESE GARDEN MRT STATION          100        100        100            100   
CHOA CHU KANG MRT STATION            75         75         75             75   
CITY HALL MRT STATION               200        200        200            200   
CLARKE QUAY MRT STATION             100        100        100            100   
CLEMENTI MRT STATION                100        100        100            100   
COMMONWEALTH MRT STATION            100        100        100            100   
DAKOTA MRT STATION                  100        100        100            100   
DHOBY GHAUT MRT STATION             300        300        300            300   
DOVER MRT STATION                   100        100        100            100   
DOWNTOWN MRT STATION                100        100        100            100   
ESPLANADE MRT STATION               100        100        100            100   
EUNOS MRT STATION                   100        100        100            100   
EXPO MRT STATION                    194        194        194            194   
FARRER PARK MRT STATION             100        100        100            100   
FARRER ROAD MRT STATION             100        100        100            100   
FORT CANNING MRT STATION            100        100        100            100   
GEYLANG BAHRU MRT STATION           100        100        100            100   
GUL CIRCLE MRT STATION               11         11         11             11   
HARBOURFRONT MRT STATION            200        200        200            200   
HAW PAR VILLA MRT STATION           100        100        100            100   
HILLVIEW MRT STATION           

In [13]:
print('There are {} uniques categories.'.format(len(venues_df['VenueCategory'].unique())))

There are 329 uniques categories.


In [14]:
# print out the list of categories
venues_df['VenueCategory'].unique()[:50]

array(['Food Court', 'Pet Store', 'Fast Food Restaurant',
       'Malay Restaurant', 'Coffee Shop', 'Park', 'Sandwich Place',
       'College Theater', 'Supermarket', 'Food Truck',
       'Frozen Yogurt Shop', 'Clothing Store', 'Movie Theater',
       'Electronics Store', 'Grocery Store', 'Dim Sum Restaurant',
       'Sushi Restaurant', 'Café', 'Shopping Mall', 'Japanese Restaurant',
       'Trail', 'Chinese Restaurant', 'Indian Restaurant', 'Post Office',
       'Ice Cream Shop', 'Snack Place', 'Night Market',
       'Convenience Store', 'Gym', 'Thai Restaurant', 'Asian Restaurant',
       'Pizza Place', 'Bookstore', 'Mobile Phone Shop',
       'Seafood Restaurant', 'Restaurant', 'Bakery',
       'Fried Chicken Joint', 'BBQ Joint', 'Soup Place',
       'Cantonese Restaurant', 'Climbing Gym', 'Kitchen Supply Store',
       'Vegetarian / Vegan Restaurant', 'Gym / Fitness Center',
       'Bubble Tea Shop', 'Performing Arts Venue', 'Soccer Field',
       'Boarding House', 'Sporting Goods 

## Step 4: With the unique categories, a few categories like "Climbing Gym", "Yoga Studio", etc. are renamed to "Gym" to consider them as fitness facilities as well.

In [36]:
venues_df.replace('Gym / Fitness Center', 'Gym', inplace = True)
venues_df.replace('Climbing Gym', 'Gym', inplace = True)
venues_df.replace('Rock Climbing Spot', 'Gym', inplace = True)
venues_df.replace('Yoga Studio', 'Gym', inplace = True)
venues_df.head()

Neighborhood  Latitude   Longitude  \
0  ADMIRALTY MRT STATION  1.440585  103.800998   
1  ADMIRALTY MRT STATION  1.440585  103.800998   
2  ADMIRALTY MRT STATION  1.440585  103.800998   
3  ADMIRALTY MRT STATION  1.440585  103.800998   
4  ADMIRALTY MRT STATION  1.440585  103.800998   

                         VenueName  VenueLatitude  VenueLongitude  \
0  Kampung Admiralty Hawker Centre       1.439939      103.800774   
1         Superpets Warehouse / HQ       1.436848      103.806059   
2                       McDonald's       1.445931      103.798101   
3                       Asia Ghani       1.437431      103.795388   
4                        Starbucks       1.439761      103.800659   

          VenueCategory  
0            Food Court  
1             Pet Store  
2  Fast Food Restaurant  
3      Malay Restaurant  
4           Coffee Shop

In [16]:
# check if the results contain "Gym"
"Gym" in venues_df['VenueCategory'].unique()

True

In [17]:
# one hot encoding
sg_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
sg_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [sg_onehot.columns[-1]] + list(sg_onehot.columns[:-1])
sg_onehot = sg_onehot[fixed_columns]

print(sg_onehot.shape)
sg_onehot.head()

(13594, 326)


Neighborhoods  Accessories Store  Airport  Airport Gate  \
0  ADMIRALTY MRT STATION                  0        0             0   
1  ADMIRALTY MRT STATION                  0        0             0   
2  ADMIRALTY MRT STATION                  0        0             0   
3  ADMIRALTY MRT STATION                  0        0             0   
4  ADMIRALTY MRT STATION                  0        0             0   

   Airport Lounge  Airport Service  American Restaurant  Aquarium  Arcade  \
0               0                0                    0         0       0   
1               0                0                    0         0       0   
2               0                0                    0         0       0   
3               0                0                    0         0       0   
4               0                0                    0         0       0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0            0           0                    0                 0   
1            0           0                    0                 0   
2            0           0                    0                 0   
3            0           0                    0                 0   
4            0           0                    0                 0   

   Athletics & Sports  Australian Restaurant  Auto Garage  BBQ Joint  \
0                   0                      0            0          0   
1                   0                      0            0          0   
2                   0                      0            0          0   
3                   0                      0            0          0   
4                   0                      0            0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0           0           0       0     0    0                 0   
1           0           0       0     0    0                 0   
2           0           0       0     0    0                 0   
3           0           0       0     0    0                 0   
4           0           0       0     0    0                 0   

   Basketball Court  Bay  Beach  Beach Bar  Bed & Breakfast  Beer Bar  \
0                 0    0      0          0                0         0   
1                 0    0      0          0                0         0   
2                 0    0      0          0                0         0   
3                 0    0      0          0                0         0   
4                 0    0      0          0                0         0   

   Beer Garden  Beer Store  Betting Shop  Big Box Store  Bike Trail  Bistro  \
0            0           0             0              0           0       0   
1            0           0             0              0           0       0   
2            0           0             0              0           0       0   
3            0           0             0              0           0       0   
4            0           0             0              0           0       0   

   Boarding House  Boat or Ferry  Bookstore  Border Crossing  \
0               0              0          0                0   
1               0              0          0                0   
2               0              0          0                0   
3               0              0          0                0   
4               0              0          0                0   

   Botanical Garden  Boutique  Bowling Alley  Bowling Green  Breakfast Spot  \
0                 0         0              0              0               0   
1                 0         0              0              0               0   
2                 0         0              0              0               0   
3                 0         0              0              0               0   
4                 0         0              0              0               0   

   Brewery  Bridge  Bubble Tea Shop  Buddhist Temple  Buffet  Building  \
0        0       0                

In [34]:
sg_grouped = sg_onehot.groupby(["Neighborhoods"]).mean().reset_index()

print(sg_grouped.shape)
sg_grouped.head()

(121, 326)


Neighborhoods  Accessories Store  Airport  Airport Gate  \
0   ADMIRALTY MRT STATION                0.0      0.0           0.0   
1    ALJUNIED MRT STATION                0.0      0.0           0.0   
2  ANG MO KIO MRT STATION                0.0      0.0           0.0   
3     BARTLEY MRT STATION                0.0      0.0           0.0   
4    BAYFRONT MRT STATION                0.0      0.0           0.0   

   Airport Lounge  Airport Service  American Restaurant  Aquarium  Arcade  \
0             0.0              0.0                  0.0       0.0     0.0   
1             0.0              0.0                  0.0       0.0     0.0   
2             0.0              0.0                  0.0       0.0     0.0   
3             0.0              0.0                  0.0       0.0     0.0   
4             0.0              0.0                  0.0       0.0     0.0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0.00        0.00                  0.0          0.012346   
1         0.00        0.00                  0.0          0.060000   
2         0.00        0.00                  0.0          0.020000   
3         0.00        0.00                  0.0          0.050000   
4         0.02        0.01                  0.0          0.000000   

   Athletics & Sports  Australian Restaurant  Auto Garage  BBQ Joint  \
0                 0.0                   0.00          0.0       0.00   
1                 0.0                   0.00          0.0       0.06   
2                 0.0                   0.01          0.0       0.02   
3                 0.0                   0.00          0.0       0.02   
4                 0.0                   0.00          0.0       0.00   

   Baby Store  Bagel Shop    Bakery  Bank   Bar  Baseball Stadium  \
0         0.0         0.0  0.012346  0.00  0.00               0.0   
1         0.0         0.0  0.000000  0.00  0.00               0.0   
2         0.0         0.0  0.030000  0.01  0.02               0.0   
3         0.0         0.0  0.000000  0.00  0.02               0.0   
4         0.0         0.0  0.000000  0.00  0.01               0.0   

   Basketball Court   Bay  Beach  Beach Bar  Bed & Breakfast  Beer Bar  \
0              0.00  0.00    0.0        0.0              0.0       0.0   
1              0.02  0.00    0.0        0.0              0.0       0.0   
2              0.00  0.00    0.0        0.0              0.0       0.0   
3              0.00  0.00    0.0        0.0              0.0       0.0   
4              0.00  0.01    0.0        0.0              0.0       0.0   

   Beer Garden  Beer Store  Betting Shop  Big Box Store  Bike Trail  Bistro  \
0         0.00         0.0           0.0            0.0         0.0    0.00   
1         0.00         0.0           0.0            0.0         0.0    0.00   
2         0.01         0.0           0.0            0.0         0.0    0.00   
3         0.00         0.0           0.0            0.0         0.0    0.00   
4         0.00         0.0           0.0            0.0         0.0    0.01   

   Boarding House  Boat or Ferry  Bookstore  Border Crossing  \
0            0.00            0.0   0.012346              0.0   
1            0.01            0.0   0.000000              0.0   
2            0.00            0.0   0.000000              0.0   
3            0.00            0.0   0.000000              0.0   
4            0.00            0.0   0.000000              0.0   

   Botanical Garden  Boutique  Bowling Alley  Bowling Green  Breakfast Spot  \
0              0.00      0.00            0.0            0.0             0.0   
1              0.00      0.00            0.0            0.0             0.0   
2              0.00      0.00            0.0            0.0             0.0   
3              0.00      0.00            0.0            0.0             0.0   
4              0.02      0.02            0.0            0.0             0.0   

   Brewery  Bridge  Bubble Tea Shop  Buddhist Temple  Buffet  Building  \
0    

In [19]:
len(sg_grouped[sg_grouped["Gym"] > 0])

103

## Step 5: Based on the data, 103 fitness facilities are identified. Now we look at where these facilities are by breaking them into 4 clusters.

In [20]:
sg_gym = sg_grouped[["Neighborhoods","Gym"]]
sg_gym.head()

Neighborhoods       Gym
0   ADMIRALTY MRT STATION  0.012346
1    ALJUNIED MRT STATION  0.050000
2  ANG MO KIO MRT STATION  0.010000
3     BARTLEY MRT STATION  0.010000
4    BAYFRONT MRT STATION  0.020000

In [21]:
# set number of clusters
kclusters = 4

sg_clustering = sg_gym.drop(["Neighborhoods"], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(sg_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 0, 1, 1, 2, 2, 2, 3, 1, 1], dtype=int32)

In [22]:
# create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.
sg_merged = sg_gym.copy()

# add clustering labels
sg_merged["Cluster Labels"] = kmeans.labels_

In [23]:
sg_merged.rename(columns={"Neighborhoods": "Neighborhood"}, inplace=True)
sg_merged.head()

Neighborhood       Gym  Cluster Labels
0   ADMIRALTY MRT STATION  0.012346               1
1    ALJUNIED MRT STATION  0.050000               0
2  ANG MO KIO MRT STATION  0.010000               1
3     BARTLEY MRT STATION  0.010000               1
4    BAYFRONT MRT STATION  0.020000               2

In [24]:
# merge sg_grouped with sg_mrt to add latitude/longitude for each neighborhood
sg_merged = sg_merged.join(sg_mrt.set_index("Neighborhood"), on="Neighborhood")

print(sg_merged.shape)
sg_merged.head() # check the last columns!

(144, 10)


Neighborhood       Gym  Cluster Labels   ID Number           X  \
0   ADMIRALTY MRT STATION  0.012346               1   12   NS10  24402.1063   
1    ALJUNIED MRT STATION  0.050000               0   16    EW9  33518.6049   
2  ANG MO KIO MRT STATION  0.010000               1   33   NS16  29807.2655   
3     BARTLEY MRT STATION  0.010000               1  153   CC12  33168.3039   
4    BAYFRONT MRT STATION  0.020000               2  115   DT16  30867.0093   

            Y  Latitude   Longitude    Line  
0  46918.1131  1.440585  103.800998     RED  
1  33190.0020  1.316433  103.882893   GREEN  
2  39105.7720  1.369933  103.849553     RED  
3  36108.7003  1.342828  103.879746  YELLOW  
4  29368.6250  1.281874  103.859073    BLUE

In [35]:
# sort the results by Cluster Labels
print(sg_merged.shape)
sg_merged.sort_values(["Cluster Labels"], inplace=True)
sg_merged.head()

(144, 10)


Neighborhood       Gym  Cluster Labels   ID Number  \
69   MARINA SOUTH PIER MRT STATION  0.040000               0  150   NS28   
55             KALLANG MRT STATION  0.040000               0   26   EW10   
77           ONE-NORTH MRT STATION  0.040000               0  108   CC23   
86             PUNGGOL MRT STATION  0.036585               0  106   NE17   
106         TELOK AYER MRT STATION  0.040000               0  190   DT18   

              X           Y  Latitude   Longitude    Line  
69   31287.7063  28203.4939  1.271337  103.862853     RED  
55   32236.6038  32643.3302  1.311489  103.871377   GREEN  
77   22896.0731  31346.4085  1.299760  103.787469  YELLOW  
86   35651.3846  42933.2694  1.404547  103.902052  PURPLE  
106  29667.6216  29414.5867  1.282290  103.848299    BLUE

# Results

The map below shows which of the MRT stations in 4 clusters. It shows which MRT stations are crowded with fitness facilities and which stations lack such facilities.

In [26]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(sg_merged['Latitude'], sg_merged['Longitude'], sg_merged['Neighborhood'], sg_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' - Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [27]:
# save the map as HTML file
map_clusters.save('map_clusters.html')

In [28]:
sg_merged.loc[sg_merged['Cluster Labels'] == 0]

Neighborhood       Gym  Cluster Labels   ID Number  \
69   MARINA SOUTH PIER MRT STATION  0.040000               0  150   NS28   
36               DOVER MRT STATION  0.040000               0   29   EW22   
101           TAMPINES MRT STATION  0.040000               0  142   DT32   
79         OUTRAM PARK MRT STATION  0.050000               0  156    NE3   
101           TAMPINES MRT STATION  0.040000               0   11    EW2   
100      TAMPINES EAST MRT STATION  0.050000               0   96   DT33   
79         OUTRAM PARK MRT STATION  0.050000               0  188   EW16   
97             STADIUM MRT STATION  0.060000               0  110    CC6   
27           CHINATOWN MRT STATION  0.045000               0  131    NE4   
27           CHINATOWN MRT STATION  0.045000               0  132   DT19   
44       GEYLANG BAHRU MRT STATION  0.040000               0   98   DT24   
94               SIMEI MRT STATION  0.040000               0   10    EW3   
82          PAYA LEBAR MRT STATION  0.050000               0    9    EW8   
82          PAYA LEBAR MRT STATION  0.050000               0  152    CC9   
34              DAKOTA MRT STATION  0.050000               0  154    CC8   
68          MARINA BAY MRT STATION  0.035000               0  129    CE2   
17            BUANGKOK MRT STATION  0.040000               0  183   NE15   
105      TANJONG PAGAR MRT STATION  0.050000               0  107   EW15   
102      TAMPINES WEST MRT STATION  0.040000               0   97   DT31   
73         MOUNTBATTEN MRT STATION  0.040000               0  118    CC7   
1             ALJUNIED MRT STATION  0.050000               0   16    EW9   
68          MARINA BAY MRT STATION  0.035000               0  128   NS27   
106         TELOK AYER MRT STATION  0.040000               0  190   DT18   
57          KENT RIDGE MRT STATION  0.040000               0  104   CC24   
86             PUNGGOL MRT STATION  0.036585               0  106   NE17   
77           ONE-NORTH MRT STATION  0.040000               0  108   CC23   
55             KALLANG MRT STATION  0.040000               0   26   EW10   

              X           Y  Latitude   Longitude    Line  
69   31287.7063  28203.4939  1.271337  103.862853     RED  
36   21914.5509  32634.1186  1.311405  103.778652   GREEN  
101  40213.0284  37463.3697  1.355077  103.943030    BLUE  
79   28696.2602  29135.3820  1.281405  103.839125  PURPLE  
101  40445.4297  37266.9909  1.353301  103.945118   GREEN  
100  41501.0749  37586.6178  1.356191  103.954600    BLUE  
79   28646.3739  29316.7727  1.279765  103.839573   GREEN  
97   32677.7526  31687.0547  1.302841  103.875339  YELLOW  
27   29124.9585  29643.4786  1.284360  103.843424  PURPLE  
27   29238.9683  29686.5359  1.284749  103.844448    BLUE  
44   32293.7994  33750.9518  1.321506  103.871890    BLUE  
94   41360.6138  36150.3959  1.343203  103.953339   GREEN  
82   34648.6773  33375.7341  1.318112  103.893045   GREEN  
82   34550.5756  33300.3402  1.317430  103.892163  YELLOW  
34   34159.2741  32299.8905  1.308383  103.888648  YELLOW  
68   30423.4296  28735.7843  1.276151  103.855089  OTHERS  
17   34655.3132  40537.2069  1.382878  103.893104  PURPLE  
105  29396.2311  28776.7576  1.276521  103.845861   GREEN  
102  39698.5276  36406.0143  1.345515  103.938408    BLUE  
73   33476.5634  32058.7418  1.306202  103.882515  YELLOW  
1    33518.6049  33190.0020  1.316433  103.882893   GREEN  
68   30368.2057  28766.3819  1.276427  103.854593     RED  
106  29667.6216  29414.5867  1.282290  103.848299    BLUE  
57   22581.5421  30650.0972  1.293463  103.784644  YELLOW  
86   35651.3846  42933.2694  1.404547  103.902052  PURPLE  
77   22896.0731  31346.4085  1.299760  103.787469  YELLOW  
55   32236.6038  32643.3302  1.311489  103.871377   GREEN

In [29]:
sg_merged.loc[sg_merged['Cluster Labels'] == 1]

Neighborhood       Gym  Cluster Labels   ID Number  \
56         KEMBANGAN MRT STATION  0.010000               1   31    EW6   
38         ESPLANADE MRT STATION  0.010000               1  125    CC3   
85         PROMENADE MRT STATION  0.010000               1  184    CC4   
51       JALAN BESAR MRT STATION  0.010000               1  114   DT22   
85         PROMENADE MRT STATION  0.010000               1  102   DT15   
84      POTONG PASIR MRT STATION  0.010000               1  176   NE10   
70         MARSILING MRT STATION  0.013158               1   36    NS8   
71         MARYMOUNT MRT STATION  0.010000               1  160   CC16   
66      LORONG CHUAN MRT STATION  0.010000               1  162   CC14   
65      LITTLE INDIA MRT STATION  0.010000               1  169    NE7   
65      LITTLE INDIA MRT STATION  0.010000               1  168   DT12   
50           HOUGANG MRT STATION  0.010000               1  177   NE14   
46      HARBOURFRONT MRT STATION  0.010000               1  101    NE1   
62     LABRADOR PARK MRT STATION  0.010526               1  130   CC27   
46      HARBOURFRONT MRT STATION  0.010000               1  101   CC29   
60             KOVAN MRT STATION  0.010000               1  119   NE13   
75    NICOLL HIGHWAY MRT STATION  0.010000               1  126    CC5   
54        KAKI BUKIT MRT STATION  0.010000               1  141   DT28   
81         PASIR RIS MRT STATION  0.010000               1   14    EW1   
78           ORCHARD MRT STATION  0.010000               1  113   NS22   
0          ADMIRALTY MRT STATION  0.012346               1   12   NS10   
99          TAI SENG MRT STATION  0.010000               1  166   CC11   
11            BISHAN MRT STATION  0.010000               1  109   CC15   
11            BISHAN MRT STATION  0.010000               1  158   NS17   
8    BEDOK RESERVOIR MRT STATION  0.010000               1  149   DT30   
114     UPPER CHANGI MRT STATION  0.010000               1  179   DT34   
107    TELOK BLANGAH MRT STATION  0.010000               1  189   CC28   
16        BRAS BASAH MRT STATION  0.010000               1   93    CC2   
18             BUGIS MRT STATION  0.010000               1  103   DT14   
18             BUGIS MRT STATION  0.010000               1  111   EW12   
21      BUKIT GOMBAK MRT STATION  0.010101               1   35    NS3   
9          BENCOOLEN MRT STATION  0.010000               1  100   DT21   
115        WOODLANDS MRT STATION  0.011050               1  181    TE2   
116  WOODLANDS SOUTH MRT STATION  0.015385               1  175    TE3   
28    CHINESE GARDEN MRT STATION  0.010000               1    2   EW25   
3            BARTLEY MRT STATION  0.010000               1  153   CC12   
90            ROCHOR MRT STATION  0.010000               1  167   DT13   
2         ANG MO KIO MRT STATION  0.010000               1   33   NS16   
32          CLEMENTI MRT STATION  0.010000               1   15   EW23   
89           REDHILL MRT STATION  0.010000               1    5   EW18   
115        WOODLANDS MRT STATION  0.011050               1   24    NS9   

              X           Y  Latitude   Longitude    Line  
56   36862.1237  33699.3568  1.321038  103.912928   GREEN  
38   30469.0566  30634.4621  1.293322  103.855498  YELLOW  
85   31080.6082  30623.0174  1.293218  103.860992  YELLOW  
51   30466.0301  31970.4313  1.305404  103.855471    BLUE  
85   31064.3414  30586.9146  1.292892  103.860846    BLUE  
84   31977.1677  34842.7289  1.331380  103.869046  PURPLE  
70   21406.6874  46025.7686  1.432514  103.774090     RED  
71   28679.3688  36758.7136  1.348707  103.839421  YELLOW  
66   31431.4054  37079.9444  1.351612  103.864143  YELLOW  
65   29817.1982  32124.8312  1.307198  103.848581  PURPLE  
65   29698.9692  32168.8611  1.303916  103.852417    BLUE  
50   34572.8759  39256.1311  1.371292  103.892364  PURPLE  
46   26678.3445  27555.0595  1.265473  103.821446  PURPLE  
62   24619.0578  28313.6298  1.272333  103.802947  YELLOW  
46   26678.

In [30]:
sg_merged.loc[sg_merged['Cluster Labels'] == 2]

Neighborhood       Gym  Cluster Labels   ID Number  \
74             NEWTON MRT STATION  0.020000               2  133   DT11   
118           YEW TEE MRT STATION  0.032258               2    6    NS5   
72             MATTAR MRT STATION  0.030000               2  138   DT25   
88      RAFFLES PLACE MRT STATION  0.020000               2  186   EW14   
108       TIONG BAHRU MRT STATION  0.020000               2  135   EW17   
113               UBI MRT STATION  0.020000               2  139   DT27   
109         TOA PAYOH MRT STATION  0.020000               2  174   NS19   
87         QUEENSTOWN MRT STATION  0.020000               2   21   EW19   
76             NOVENA MRT STATION  0.030000               2  147   NS20   
80      PASIR PANJANG MRT STATION  0.020000               2  155   CC26   
98            STEVENS MRT STATION  0.020000               2  170   DT10   
96           SOMERSET MRT STATION  0.020000               2  112   NS23   
95       SIXTH AVENUE MRT STATION  0.031579               2  173    DT7   
92           SENGKANG MRT STATION  0.030000               2  116   NE16   
83            PIONEER MRT STATION  0.020000               2    7   EW28   
74             NEWTON MRT STATION  0.020000               2  191   NS21   
88      RAFFLES PLACE MRT STATION  0.020000               2  186   NS26   
59   KING ALBERT PARK MRT STATION  0.020000               2  137    DT6   
67         MACPHERSON MRT STATION  0.030000               2   94   CC10   
31        CLARKE QUAY MRT STATION  0.020000               2  157    NE5   
30          CITY HALL MRT STATION  0.020000               2   95   NS25   
67         MACPHERSON MRT STATION  0.030000               2  105   DT26   
30          CITY HALL MRT STATION  0.020000               2   95   EW13   
29      CHOA CHU KANG MRT STATION  0.026667               2   19    NS4   
25             CASHEW MRT STATION  0.020000               2  172    DT2   
24          CALDECOTT MRT STATION  0.020000               2  127   CC17   
23        BUONA VISTA MRT STATION  0.020000               2  124   CC22   
33       COMMONWEALTH MRT STATION  0.020000               2    8   EW20   
23        BUONA VISTA MRT STATION  0.020000               2   28   EW21   
15           BRADDELL MRT STATION  0.020000               2  182   NS18   
13           BOON LAY MRT STATION  0.020833               2   34   EW27   
12          BOON KENG MRT STATION  0.020000               2  178    NE9   
10          BENDEMEER MRT STATION  0.020000               2   99   DT23   
6               BEDOK MRT STATION  0.020000               2   37    EW5   
5        BEAUTY WORLD MRT STATION  0.023256               2  140    DT5   
4            BAYFRONT MRT STATION  0.020000               2  115    CE1   
4            BAYFRONT MRT STATION  0.020000               2  115   DT16   
22      BUKIT PANJANG MRT STATION  0.020619               2  171    DT1   
35        DHOBY GHAUT MRT STATION  0.030000               2  144   NS24   
120            YISHUN MRT STATION  0.020000               2   13   NS13   
35        DHOBY GHAUT MRT STATION  0.030000               2  165    CC1   
53        JURONG EAST MRT STATION  0.021505               2   20    NS1   
35        DHOBY GHAUT MRT STATION  0.030000               2  164    NE6   
49    HOLLAND VILLAGE MRT STATION  0.020000               2  187   CC21   
47      HAW PAR VILLA MRT STATION  0.030000               2  161   CC25   
58             KHATIB MRT STATION  0.030000               2    3   NS14   
53        JURONG EAST MRT STATION  0.021505               2   20   EW24   
119      YIO CHU KANG MRT STATION  0.023810               2   41   NS15   
43       FORT CANNING MRT STATION  0.030000               2  120   DT20   
63           LAKESIDE MRT STATION  0.020000               2   39   EW26   
64           LAVENDER MRT STATION  0.030000               2  163   EW11   
39              EUNOS MRT STATION  0.030000               2    1    EW7   
37           DOWNTOWN MRT STATION  0.030000   

In [31]:
sg_merged.loc[sg_merged['Cluster Labels'] == 3]

Neighborhood    Gym  Cluster Labels   ID Number  \
7        BEDOK NORTH MRT STATION  0.000               3  143   DT29   
61            KRANJI MRT STATION  0.000               3    4    NS7   
117        WOODLEIGH MRT STATION  0.000               3  146   NE11   
112   TUAS WEST ROAD MRT STATION  0.000               3   27   EW32   
111        TUAS LINK MRT STATION  0.000               3   22   EW33   
110    TUAS CRESCENT MRT STATION  0.000               3   25   EW31   
14   BOTANIC GARDENS MRT STATION  0.000               3  159    DT9   
103      TAN KAH KEE MRT STATION  0.000               3  134    DT8   
52          JOO KOON MRT STATION  0.000               3   38   EW29   
104      TANAH MERAH MRT STATION  0.000               3   40    EW4   
19       BUKIT BATOK MRT STATION  0.000               3   32    NS2   
20       BUKIT BROWN MRT STATION  0.000               3  117   CC18   
48          HILLVIEW MRT STATION  0.000               3  136    DT3   
45        GUL CIRCLE MRT STATION  0.000               3   23   EW30   
26    CHANGI AIRPORT MRT STATION  0.000               3  180    CG2   
42       FARRER ROAD MRT STATION  0.000               3  121   CC20   
40              EXPO MRT STATION  0.000               3  122   DT35   
93         SERANGOON MRT STATION  0.005               3  145   CC13   
40              EXPO MRT STATION  0.000               3   18    CG1   
91         SEMBAWANG MRT STATION  0.000               3   30   NS11   
14   BOTANIC GARDENS MRT STATION  0.000               3  185   CC19   
93         SERANGOON MRT STATION  0.005               3  148   NE12   

              X           Y  Latitude   Longitude    Line  
7    37421.7438  35214.6804  1.334742  103.917955    BLUE  
61   20081.6974  45214.5479  1.425178  103.762187     RED  
117  32173.3186  35706.3794  1.339190  103.870808  PURPLE  
112   6442.6401  34689.7988  1.329989  103.639666   GREEN  
111   6138.3115  35848.0658  1.340464  103.636932   GREEN  
110   7495.5310  33698.6818  1.321027  103.649124   GREEN  
14   26087.2663  33852.4484  1.322424  103.816136    BLUE  
103  25106.8667  34234.9609  1.325883  103.807329    BLUE  
52   10756.3543  34438.2764  1.327717  103.678417   GREEN  
104  40578.8616  34379.4144  1.327187  103.946316   GREEN  
19   18679.3541  36795.0141  1.349034  103.749590     RED  
20   27707.4166  35102.4802  1.333729  103.830690  YELLOW  
48   20666.0676  38266.7899  1.362345  103.767437    BLUE  
45    8772.8552  33530.3675  1.319505  103.660599   GREEN  
26   45254.8614  37711.0150  1.357315  103.988321  OTHERS  
42   25136.2263  33309.1617  1.317511  103.807593  YELLOW  
40   42362.7103  35285.6654  1.335382  103.962341    BLUE  
93   32345.7724  36967.5123  1.350595  103.872357  YELLOW  
40   42270.6232  35193.6223  1.334550  103.961513  OTHERS  
91   26522.9883  47854.1998  1.449051  103.820050     RED  
14   25959.4046  33817.7525  1.322110  103.814988  YELLOW  
93   32480.0726  36869.3924  1.349708  103.873564  PURPLE

# Discussion

Based on the analysis, Cluster 0 is the most crowded with fitness facilities while Cluster 3 lacks fitness facilities. This shows that Central Singapore and East Singapore are crowded with fitness facilities already and I should avoid those areas. While the western part of Singapore might have opportunity for me to set up a Gym.

# Conclsuion

With the help of Foursquare, I am able to get a quick sense of where the opportunities for a Gym are. It seems like Western Singapore, such as Jurong East, Chinese Gardens MRT Stations are possible areas for a new Gym.